### PR TITLE
TGUI Hive Status Revived feat. Compact (Classic) Mode

### PR DESCRIPTION
**Pull request by Wayland-Smithy**
## About The Pull Request

Continuation of #10273 with the following changes:
- Clicking on a Xeno's name in the list will track it.
- Added a "Compact Mode" for the counts of each caste that is very reminiscent of the old Hive Status text.

image
image
image

Closes: #10399

## Why It's Good For The Game

TGUI good.

## Changelog
:cl: KornFlaks (RIP) & Wayland-Smithy
add: Xeno hive status is now a constantly updating TGUI page.
add: Xeno hivemind players can now select a xeno to jump to using the hive status UI overwatch button.
add: Evolving now closes all UI pages open.
add: Xeno purchased primordial tiers now display as a green [PRIMORDIAL] next to the tier population in the hive status UI.
imageadd: Ported CM hive status SSD icon.
qol: Evolution is displayed and can now be accessed through the hive status UI.
qol: Xeno overwatch selection now done through hive status UI. Action button no longer visible.
qol: Xeno overwatch can now be canceled without having to move by selecting the same xeno to overwatch again.
qol: Queen leadership toggle now done through hive status UI. Action button no longer visible.
qol: Queen transfer plasma now possible through hive status UI.
qol: Xeno minimap icons now display next to corresponding xeno names in hive status UI.
qol: Xeno collapse timers now display in hive status UI instead of game tab.
qol: Hive status UI can sort xeno lists by clicking on the column headers. Sorting is done by xeno tier, health points, plasma remaining, and location by alphabetical order to the first 5 letters.
fix: Queen transfer plasma message now only visible to xenos.
fix: Ghost orbiting code now properly returns an accurate reference to the atom it is orbiting.
refactor: Hive upgrades purchasing UI relocated to separate datum.
refactor: Distress mode collapse timers are now a define.
del: Old HTML hive status page now lives at your grandparents' farm upstate.
del: Queen order, the unused direct to xeno version, has been removed.
del: Removed a vestigial time check that checked if a distress beacon could be launched at 10 minutes into a round. Xenos can not hijack before 30 minutes into a round.
/:cl:

### CODE REVIEW

**Review by Kuro020 — APPROVED** [review]

**Comment by Al-1ce** [comment]
Does this affect the compass utility? Will that be removed as it were with the last TGUI xeno-status?

>Compact Mode

I'd really appreciate just a 'classic' mode that reverts to the old style. It's a lot more utilitarian and much more familiar after three years of using it.

**Comment by Wayland-Smithy** [comment]
> Does this affect the compass utility? Will that be removed as it were with the last TGUI xeno-status?

The compass loses no functionality with this iteration because just like before xenos and structures can be selected as the compass/tracking target.

> > Compact Mode
> 
> I'd really appreciate just a 'classic' mode that reverts to the old style. It's a lot more utilitarian and much more familiar after three years of using it.

Unfortunately so much under the hood has changed that it would be non trivial to keep the entire old UI.

I believe it is a good compromise in the name of progress to have the "Hive Population" section displayed in the old style so by using that and the above section you can find all the same info as before and more with less screen real estate like psy points, burrowed count, and caste totals.

File: code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm [hunk 1]
```diff
@@ -226,13 +226,16 @@
...
-		if(!check_weeds(get_turf(xeno), TRUE))
-			to_chat(src, span_warning("They are not near any weeds we can jump to."))
-			return
-		if(!(status_flags & INCORPOREAL))
-			start_teleport(get_turf(xeno))
-			return
-		abstract_move(get_turf(xeno))
+		jump(xeno)
+
+/mob/living/carbon/xenomorph/hivemind/proc/jump(mob/living/carbon/xenomorph/xeno)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm:231]: autodoc

File: code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm [hunk 2]
```diff
@@ -226,13 +226,16 @@
...
+	if(!check_weeds(get_turf(xeno), TRUE))
+		src.balloon_alert(src, "No nearby weeds")
+		return
+	if(!(status_flags & INCORPOREAL))
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm:235]: src.proc

File: code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm [hunk 3]
```diff
@@ -282,60 +264,57 @@
...
+/datum/action/xeno_action/set_xeno_lead/give_action(mob/living/L)
+	. = ..()
+	RegisterSignal(L, COMSIG_XENOMORPH_LEADERSHIP, .proc/try_use_action)
 
-	select_xeno_leader()
+/datum/action/xeno_action/set_xeno_lead/remove_action(mob/living/L)
+	. = ..()
+	UnregisterSignal(L, COMSIG_XENOMORPH_LEADERSHIP)
 
+/datum/action/xeno_action/set_xeno_lead/proc/try_use_action(datum/source, mob/living/carbon/xenomorph/target)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm:280]: autodoc

File: code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm [hunk 4]
```diff
@@ -384,45 +362,29 @@
...
 
+/datum/action/xeno_action/activable/queen_give_plasma/give_action(mob/living/L)
+	. = ..()
+	RegisterSignal(L, COMSIG_XENOMORPH_QUEEN_PLASMA, .proc/try_use_ability)
+
+/datum/action/xeno_action/activable/queen_give_plasma/remove_action(mob/living/L)
+	. = ..()
+	UnregisterSignal(L, COMSIG_XENOMORPH_QUEEN_PLASMA)
+
+/datum/action/xeno_action/activable/queen_give_plasma/proc/try_use_ability(datum/source, mob/living/carbon/xenomorph/target)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/queen/abilities_queen.dm:373]: autodoc

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 5]
```diff
@@ -52,60 +49,215 @@
...
+	.["user_tracked"] = isxeno(user) && !isnull(xeno_user.tracked) ? REF(xeno_user.tracked) : ""
+
+	.["user_show_empty"] = isxeno(user) ? xeno_user.show_empty_castes : 0
+	.["user_show_compact"] = isxeno(user) ? xeno_user.compact_mode : 0
+	.["user_show_general"] = isxeno(user) ? xeno_user.show_general : 0
+	.["user_show_population"] = isxeno(user) ? xeno_user.show_population : 0
+	.["user_show_xeno_list"] = isxeno(user) ? xeno_user.show_xeno_list : 0
+	.["user_show_structures"] = isxeno(user) ? xeno_user.show_structures : 0
+
+/datum/hive_status/proc/get_structure_packet(obj/structure/xeno/struct)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm:153]: autodoc

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 6]
```diff
@@ -52,60 +49,215 @@
...
+
+/// Returns the string location of the xeno.
+/datum/hive_status/proc/get_xeno_location(atom/xeno)
+	. = "Unknown"
+	if(is_centcom_level(xeno.z))
+		return
 
+	var/area/A = get_area(xeno)
+	if(A)
+		. = A.name
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm:260]: Please use a better varname than A

File: code/modules/mob/living/carbon/xenomorph/xeno_defines.dm [hunk 7]
```diff
@@ -257,6 +257,18 @@
...
+	///Cached state displaying the caste totals in compact mode and hide caste icons.
+	var/compact_mode = FALSE
+	///Cached state for displaying general tab in the hive status UI.
+	var/show_general = TRUE
+	///Cached state for displaying hive population tab in the hive status UI.
+	var/show_population = TRUE
+	///Cached state for displaying xeno list tab in the hive status UI.
+	var/show_xeno_list = TRUE
+	///Cached state for displaying structures tab in the hive status UI.
+	var/show_structures = TRUE
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:271]: Why are these not bitflags?

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Comment by Al-1ce** [comment]
>The compass loses no functionality with this iteration because just like before xenos and structures can be selected as the compass/tracking target.


We can't track other xenos with the compass anymore. That's definitely a loss of functionality and for no apparent reason.

**Comment by Al-1ce** [comment]
excellent stealthnerf. compass my beloved

**Pull request merged**